result_set.go

```diff
@@ -356,31 +362,140 @@ func (res ResultSet) scanRow(row *nebula.Row, colNames []string, t reflect.Type)
-		case reflect.Float32:
-			val.Field(fIdx).SetFloat(rowVal.GetFVal())
-		case reflect.Float64:
-			val.Field(fIdx).SetFloat(rowVal.GetFVal())
-		case reflect.String:
-			val.Field(fIdx).SetString(string(rowVal.GetSVal()))
-		default:
-			return val, errors.New("scan: not support type")
+		if f.Type.Kind() == reflect.Slice {
+			list := rowVal.GetLVal()
+			err := scanListCol(list.Values, structVal.Field(fIdx), f.Type)
+			if err != nil {
+				return result, err
+			}
+		} else {
+			err := scanPrimitiveCol(rowVal, structVal.Field(fIdx), f.Type.Kind())
+			if err != nil {
+				return result, err
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func scanListCol(vals []*nebula.Value, listVal reflect.Value, sliceType reflect.Type) error {
+	switch sliceType.Elem().Kind() {
+	case reflect.Struct:
+		var listCol = reflect.MakeSlice(sliceType, 0, len(vals))
+		for _, val := range vals {
+			ele := reflect.New(sliceType.Elem()).Elem()
+			err := scanStructField(val, ele, sliceType.Elem())
+			if err != nil {
+				return err
+			}
+			listCol = reflect.Append(listCol, ele)
+		}
+		listVal.Set(listCol)
+	case reflect.Ptr:
+		var listCol = reflect.MakeSlice(sliceType, 0, len(vals))
+		for _, val := range vals {
+			ele := reflect.New(sliceType.Elem().Elem())
+			err := scanStructField(val, reflect.Indirect(ele), sliceType.Elem().Elem())
+			if err != nil {
+				return err
+			}
+			listCol = reflect.Append(listCol, ele)
+		}
+		listVal.Set(listCol)
+	default:
+		return errors.New("scan: not support list type")
+	}
+
+	return nil
+}
+
+func scanStructField(val *nebula.Value, eleVal reflect.Value, eleType reflect.Type) error {
+	vertex := val.GetVVal()
+	if vertex != nil {
+		tags := vertex.GetTags()
+		vid := vertex.GetVid()
+
+		if len(tags) != 0 {
+			tag := tags[0]
+
+			props := tag.GetProps()
+			props["_vid"] = vid
+			tagName := tag.GetName()
+			props["_tag_name"] = &nebula.Value{SVal: tagName}
+
+			err := scanValFromProps(props, eleVal, eleType)
+			if err != nil {
+				return err
+			}
+			return nil
 		}
+		// no tags, continue
 	}
 
-	return val, nil
+	edge := val.GetEVal()
+	if edge != nil {
+		props := edge.GetProps()
+
+		src := edge.GetSrc()
+		dst := edge.GetDst()
+		name := edge.GetName()
+		props["_src"] = src
+		props["_dst"] = dst
+		props["_name"] = &nebula.Value{SVal: name}
+
+		err := scanValFromProps(props, eleVal, eleType)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func scanValFromProps(props map[string]*nebula.Value, val reflect.Value, tpe reflect.Type) error {
+	for fIdx := 0; fIdx < tpe.NumField(); fIdx++ {
+		f := tpe.Field(fIdx)
+		n := f.Tag.Get("nebula")
+		v, ok := props[n]
+		if !ok {
+			continue
+		}
+		err := scanPrimitiveCol(v, val.Field(fIdx), f.Type.Kind())
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func scanPrimitiveCol(rowVal *nebula.Value, val reflect.Value, kind reflect.Kind) error {
+	switch kind {
+	case reflect.Bool:
```

Comment:
when rowVal is Null type, how to set value for reflect.Value?

Reply:
need to process these two types.
![image](https://github.com/user-attachments/assets/ef25c64d-7792-4c56-b494-0ff80b6bb494)

Reply:
should we just skip to set the value if empty or null? I mean

```

func scanPrimitiveCol(rowVal *nebula.Value, val reflect.Value, kind reflect.Kind) error {
    if rowVal.IsSetNVal() {
        return nil
    }
    ...
}

```

But, is there a easy way to do `IsEmpty` on the type `*nebula.Value`?

Reply:
May I move the method `GetType` from `ValueWrapper` to `*Value`?

Reply:
BTW, what is the situation will output `empty` or `null`? I running these codes in my server for months already.

Reply:
can we just put the empty or null value into reflect.Value

Reply:
you means

```go

func scanPrimitiveCol(rowVal *nebula.Value, val reflect.Value, kind reflect.Kind) error {
    w := ValueWrapper{value: rowVal}
    if w.IsNull() || w.IsEmpty() {
        val.SetZero()
        return nil
    }

    ......
}

```

Reply:
Emm, will cause a bit performance cost but not a problem.
And, a bit ugly 😂

Reply:
ok, tests passed